lib/src/css_mirroring/bidi_css_gen.dart

```diff
@@ -0,0 +1,334 @@
+
+    String getText(TextEditTransaction t) => (t.commit()..build('')).text;
+
+    return [
+      getText(orientationNeutral),
+      getText(orientationSpecific),
+      getText(flippedOrientationSpecific)
+    ].join('\n');
+  }
+
+  /// Makes transaction from input string.
+  static TextEditTransaction _makeTransaction(String inputCss, String url) =>
+      new TextEditTransaction(inputCss, new SourceFile(inputCss, url: url));
+
+  /// Takes transaction to edit, the retention mode which defines which part to retain and the direction of the output css.
+  /// In case rulesets it drops declarations in them and if all the declaration in it have to be removed, it removes the rule itself.
+  /// In case Directives, it edits rulesets in them and if all the rulesets have to be removed, it removes Directive Itself
+  _editTransaction(TextEditTransaction trans, RetentionMode mode, Direction targetDirection) {
+
+    /// Iterate over topLevels.
+    for (int iTopLevel = 0; iTopLevel < _originalTopLevels.length; iTopLevel++) {
+      var originalTopLevel = _originalTopLevels[iTopLevel];
+      var flippedTopLevel = _flippedTopLevels[iTopLevel];
+
+      if (originalTopLevel is RuleSet && flippedTopLevel is RuleSet) {
+        _editRuleSet(trans, mode, targetDirection, iTopLevel);
+      }
+      else if(originalTopLevel.runtimeType == flippedTopLevel.runtimeType && (originalTopLevel is MediaDirective || originalTopLevel is HostDirective)) {
+        _editDirectives(trans, originalTopLevel, flippedTopLevel, mode, targetDirection);
+      }
+      else if(originalTopLevel.runtimeType == flippedTopLevel.runtimeType && _isDirectionIndependent(originalTopLevel)) {
+        if (mode != RetentionMode.keepBidiNeutral) {
+          _removeRuleSet(trans,  mode == RetentionMode.keepFlippedBidiSpecific ? _flippedTopLevels : _originalTopLevels, iTopLevel);
+        }
+      }
+      else {
+        checkState(originalTopLevel.runtimeType == flippedTopLevel.runtimeType);
+      }
+    }
+  }
+
+  /// Edit the topLevel Ruleset.
+  /// It takes transaction, the topLevels of original and flipped css, Retantion mode, Direction of output css, the index of current topLevel and end of parent of topLevel.
+   _editRuleSet(TextEditTransaction trans, RetentionMode mode, Direction targetDirection, int iTopLevel) {
+     var removals = new _PendingRemovals(trans);
+     var usedTopLevels = mode == RetentionMode.keepFlippedBidiSpecific ? _flippedTopLevels : _originalTopLevels;
+    _storeRemovableDeclarations(removals, _originalTopLevels, _flippedTopLevels, mode, iTopLevel);
+    if (_isRuleRemovable(removals, _originalTopLevels[iTopLevel])) {
+      _removeRuleSet(trans, usedTopLevels, iTopLevel);
+    }
+    else {
+      removals.commitRemovals();
+      /// Add direction attribute to RuleId for direction-specific RuleSet.
+      if (mode != RetentionMode.keepBidiNeutral) {
```

Comment:
How about passing a null targetDirection and test if targetDirection != null here instead? It would be easier to understand the getOutputCss logic (passing directions null, native, flipped(native)).

And... now it's clear we could even remove that mode altogether and just pass the right top-levels to this method, which signature could look like:

    String _buildCssFragment(String css, List<TreeNode> topLevels, Direction directionToAnnotateRulesWith)